New: Add plasticity to grain size model

jdannberg opened the pull request:
This PR adds plasticity (Drucker-Prager yield criterion) to the grain size material model. Grain size evolution is not (yet) affected by plasticity. Below I show images of the tests I made. The `gs_drucker_prager_extension` test is based on the existing `drucker_prager_extension` test, which uses the `drucker prager` material model. However, since yielding is implemented slightly different in that model (there is not background viscous viscosity, which affects the strain rate in the first iteration, and instead of a cutoff between min and max viscosity, it uses a harmonic average), the results are only qualitatively the same. They became more similar when I used more iterations, and when I hacked in the harmonic averaging, just to compare (see images below):

(Figure below) Image of the `gs_drucker_prager_extension` output (top) vs the `drucker_prager_extension` output (bottom) after 50 iterations (rather than the 10 being done in the test). 
![drucker_prager2](https://github.com/geodynamics/aspect/assets/7517347/30d56292-271e-499b-ad84-0f23eaad0a4e)


(Figure below) Image of the `gs_drucker_prager_extension` output with harmonic averaging rather than viscosity cutoff (top) vs the `drucker_prager_extension` output (bottom) after 50 iterations (rather than the 10 being done in the test). 
![drucker_prager](https://github.com/geodynamics/aspect/assets/7517347/d39522d6-3f7d-4c8b-906a-fad34566b570)


Since I was not happy with that, I added the test `gs_drucker_prager` test. In this test, pressure increases linearly along the x axis and strain rate increases linearly along the y axis. This means we know where the yield envelope should be. 
![yielding](https://github.com/geodynamics/aspect/assets/7517347/1bc4235d-5862-488d-8ffb-6e52d49690e3)

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

naliboff reviewed (changes requested):
@jdannberg - Thanks for the contribution and this PR is in great shape! My only suggestion at this stage is to add a third test based on one of the visco plastic model test suites for a single composition + plasticity. I think this material model can be configured to reproduce exactly the same output, which would be a nice confirmation.

jdannberg commented:
@naliboff I have added another test. I modified the `tests/visco_plastic_yield.prm` file to include more outputs (the viscosity) and a statistics file. The results are now exactly the same (other than the grain size model needing a compositional field).

naliboff reviewed (approved):
Thank you for adding the additional test, good to go!

naliboff commented:
@jdannberg - Some of the tests need to be updated, but after that is done this is ready to be merged.
